cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -287,18 +277,188 @@ void ProcessListScalar(const ScalarType& scalar, ValidFunc&& valid_func,
+    valid_func(list_length, scalar_validity_buf, scalar_data_buf_1, scalar_data_buf_2);
   } else { // data.scalar->is_valid == false
     null_func();
   }
 }
 
+// A template to know how much more to allocate for the encoding of a list type for a
+// batch.
+// get_list_length must take
+// - the starting offset of the list
+// - the ending offset of the list
+// - the start of the validity buffer of the nested-type
+// - the start of the nested-type's buffers[1]
+// - the start of the nested-type's buffers[2] or NULLPTR if it does not exist
+// - returns: the amount of bytes to allocate for the list; get_list_length(0, 0, NULLPTR,
+//   NULLPTR, NULLPTR) must return the amount of space to allocate for a null list
+template<typename T, typename LengthFunc>
+void AddListLength(const ExecValue& data, int64_t batch_length, int32_t* lengths,
+                  LengthFunc&& get_list_length) {
+  using Offset = typename T::offset_type;
+  using ScalarType = typename TypeTraits<T>::ScalarType;
+
+  const int32_t null_list_length = get_list_length(0, 0, NULLPTR, NULLPTR, NULLPTR);
+  if (data.is_array()) {
+    const ArraySpan& arr = data.array;
+    const uint8_t* child_validity_buffer = arr.child_data[0].GetValues<uint8_t>(0);
+    const uint8_t* child_data_buffer_1 = arr.child_data[0].GetValues<uint8_t>(1);
+    const uint8_t* child_data_buffer_2 = arr.child_data[0].GetValues<uint8_t>(2);
+
+    int64_t i = 0;
+    IterateListArray<T>(
+      arr,
+      [&](Offset list_begin, Offset list_end) {
+        lengths[i++] += get_list_length(list_begin, list_end,
+          child_validity_buffer, child_data_buffer_1, child_data_buffer_2);
+      },
+      [&] { lengths[i++] += null_list_length; });
+  } else {
+    int32_t buffer_size;
+    ProcessListScalar<T>(
+      data.scalar_as<ScalarType>(),
+      [&](int64_t list_length, const uint8_t* child_validity_buf,
+        const uint8_t* child_data_buf_1, const uint8_t* child_data_buf_2) {
+        buffer_size = get_list_length(0, list_length,
+          child_validity_buf, child_data_buf_1, child_data_buf_2);
+      },
+      [&](){
+        buffer_size = null_list_length;
+      });
+    for (int64_t i = 0; i < batch_length; i++) {
+      lengths[i] += buffer_size;
+    }
+  }
+}
+
+// A template to encode a list type for a batch.
+// get_list_length must take (same as AddListLength)
+// - the starting offset of the list
+// - the ending offset of the list
+// - the start of the validity buffer of the nested-type
+// - the start of the nested-type's buffers[1]
+// - the start of the nested-type's buffers[2] or NULLPTR if it does not exist
+// - returns: the amount of bytes to allocate for the list; getListLength(0, 0, NULLPTR,
+//   NULLPTR, NULLPTR) must return the amount of space to allocate for a null list
+// encode_element and encode_null_element encodes elements of the lists
+// - take j, representing the element's position in the child array
```

Comment:
I'll replace it with `value array` as it is where `j` points to. 

It is pretty confusing, since for decode it refers to the j-th element in the i-th list, but here it is the j-th element of the value array.

Reply:
ah i see. makes sense